sdk/metric/internal/exemplar/reservoir.go

```diff
@@ -0,0 +1,51 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exemplar // import "go.opentelemetry.io/otel/sdk/metric/internal/exemplar"
+
+import (
+	"context"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+)
+
+// Reservoir holds the sampled exemplar of measurements made.
+type Reservoir[N int64 | float64] interface {
+	// Offer accepts the parameters associated with a measurement. The
+	// parameters will be stored as an exemplar if the Reservoir decides to
+	// sample the measurement.
+	//
+	// The passed ctx needs to contain any baggage or span that were active
+	// when the measurement was made. This information may be used by the
+	// Reservoir in making a sampling decision.
+	//
+	// The time t is the time when the measurement was made. The val and attr
+	// parameters are the value and dropped (filtered) attributes of the
+	// measurement respectively.
+	Offer(ctx context.Context, t time.Time, val N, attr []attribute.KeyValue)
```

Comment:
Do implementations need to make a copy of the attributes?  Are they allowed to modify them?

Reply:
Same resolution as https://github.com/open-telemetry/opentelemetry-go/pull/4835#discussion_r1463702803

Reply:
I don't think implementations need to make a copy here. These are passed so they will end up in the exported `metricdata.Exemplar`. They will use the same underlying array as the attributes the user originally passed:

https://github.com/open-telemetry/opentelemetry-go/pull/4455/files#diff-2c6f5806e27a5ab81eeba6f9d6fff1a80ac35188c8e22deea1c97a77fdb04634R75-R84

Since the user passes a `Set` there is no modification that the user can make to this data, and the SDK also does not modify this data.

Again, I think when we export this interface we may need to document this better there, but for our internal use I'm not sure it is needed. Do you?

Reply:
No.  I had assumed this was temporarily internal, and would be the user-facing interface eventually.